src/player/Player.php

```diff
@@ -1685,15 +1685,24 @@ public function pickBlock(Vector3 $pos, bool $addTileNBT) : bool{
 			return true;
 		}
 
-		$item = $block->getPickedItem($addTileNBT);
+		$pickedItem = $block->getPickedItem($addTileNBT);
 
-		$ev = new PlayerBlockPickEvent($this, $block, $item);
-		$existingSlot = $this->inventory->first($item);
-		if($existingSlot === -1 && $this->hasFiniteResources()){
-			$ev->cancel();
-		}
+		$ev = new PlayerBlockPickEvent($this, $block, $pickedItem);
+		$checkExistingSlot = function(Item $item) use ($ev) : int{
+			$existingSlot = $this->inventory->first($item);
+			if($existingSlot === -1 && $this->hasFiniteResources()){
+				$ev->cancel();
+			}
+			return $existingSlot;
+		};
+		$existingSlot = $checkExistingSlot($pickedItem);
 		$ev->call();
 
+		$resultItem = $ev->getResultItem();
+		if(!$resultItem->equalsExact($pickedItem)){
```

Comment:
Why rechecking again when you can do it all at once after the call? The item will be the same if there has been no change, or it will be modified directly.